File: scripts/analyze-csv-data.php
<?php
/**
 * CSV Data Analysis Script for Disability Services
 * 
 * This script analyzes disability service facility CSV data and generates
 * a comprehensive data summary JSON file for article creation.
 * 
 * Usage: php analyze-csv-data.php <csv_directory> [output_file] [baseline_file]
 * 
 * @param csv_directory Directory containing CSV files (e.g., resources/csv/202203)
 * @param output_file Optional output JSON file path (default: data-summary.json in csv_directory)
 * @param baseline_file Optional baseline JSON file for comparison (e.g., previous period data)
 */

if ($argc < 2) {
    echo "Usage: php analyze-csv-data.php <csv_directory> [output_file] [baseline_file]\n";
    echo "Example: php analyze-csv-data.php resources/csv/202203\n";
    echo "Example: php analyze-csv-data.php resources/csv/202203 custom-output.json resources/articles/202111/data-summary.json\n";
    exit(1);
}

$csvDirectory = $argv[1];
$outputFile = $argv[2] ?? $csvDirectory . '/data-summary.json';
$baselineFile = $argv[3] ?? null;

// Load baseline data for comparison if provided
$baselineData = null;
if ($baselineFile && file_exists($baselineFile)) {
    $baselineData = json_decode(file_get_contents($baselineFile), true);
}

// Service code to name mapping (from config/service.php)
$serviceMap = [
    '11' => '居宅介護',
    '12' => '重度訪問介護',
    '13' => '行動援護',
    '14' => '重度障害者等包括支援',
    '15' => '同行援護',
    '21' => '療養介護',
    '22' => '生活介護',
    '24' => '短期入所',
    '32' => '施設入所支援',
    '33' => '共同生活援助',
    '34' => '宿泊型自立訓練',
    '41' => '自立訓練(機能訓練)',
    '42' => '自立訓練(生活訓練)',
    '45' => '就労継続支援A型',
    '46' => '就労継続支援B型',
    '52' => '計画相談支援',
    '53' => '地域相談支援(地域移行)',
    '54' => '地域相談支援(地域定着)',
    '60' => '就労移行支援',
    '61' => '自立生活援助',
    '62' => '就労定着支援',
    '63' => '児童発達支援',
    '64' => '医療型児童発達支援',
    '65' => '放課後等デイサービス',
    '66' => '居宅訪問型児童発達支援',
    '67' => '保育所等訪問支援',
    '68' => '福祉型障害児入所施設',
    '69' => '医療型障害児入所施設',
    '70' => '障害児相談支援'
];

// Prefecture code to name mapping
$prefectureMap = [
    '01' => '北海道', '02' => '青森県', '03' => '岩手県', '04' => '宮城県', '05' => '秋田県',
    '06' => '山形県', '07' => '福島県', '08' => '茨城県', '09' => '栃木県', '10' => '群馬県',
    '11' => '埼玉県', '12' => '千葉県', '13' => '東京都', '14' => '神奈川県', '15' => '新潟県',
    '16' => '富山県', '17' => '石川県', '18' => '福井県', '19' => '山梨県', '20' => '長野県',
    '21' => '岐阜県', '22' => '静岡県', '23' => '愛知県', '24' => '三重県', '25' => '滋賀県',
    '26' => '京都府', '27' => '大阪府', '28' => '兵庫県', '29' => '奈良県', '30' => '和歌山県',
    '31' => '鳥取県', '32' => '島根県', '33' => '岡山県', '34' => '広島県', '35' => '山口県',
    '36' => '徳島県', '37' => '香川県', '38' => '愛媛県', '39' => '高知県', '40' => '福岡県',
    '41' => '佐賀県', '42' => '長崎県', '43' => '熊本県', '44' => '大分県', '45' => '宮崎県',
    '46' => '鹿児島県', '47' => '沖縄県'
];

echo "Analyzing CSV data in: $csvDirectory\n";

// Find all CSV files
$csvFiles = glob($csvDirectory . '/*.csv');
if (empty($csvFiles)) {
    die("No CSV files found in directory: $csvDirectory\n");
}

echo "Found " . count($csvFiles) . " CSV files\n";

$totalFacilities = 0;
$totalCapacity = 0;
$serviceStats = [];
$prefectureStats = [];
$companyStats = [];

foreach ($csvFiles as $csvFile) {
    $filename = basename($csvFile);
    
    // Extract service code from filename (e.g., csvdownload011.csv -> 11)
    if (preg_match('/csvdownload(\d+)\.csv/', $filename, $matches)) {
        $serviceCode = ltrim($matches[1], '0');
        // Fix for service 070 -> should be 70, not 7
        if ($matches[1] === '070') {
            $serviceCode = '70';
        }
        $serviceName = $serviceMap[$serviceCode] ?? "Unknown Service ($serviceCode)";
        
        echo "Processing: $filename (Service: $serviceName)\n";
        
        $handle = fopen($csvFile, 'r');
        if ($handle === false) {
            echo "Error: Cannot open file $csvFile\n";
            continue;
        }
        
        // Read header
        $header = fgetcsv($handle);
        if ($header === false) {
            echo "Error: Cannot read header from $csvFile\n";
            fclose($handle);
            continue;
        }
        
        $facilityCount = 0;
        $capacitySum = 0;
        $prefCounts = [];
        $companyCounts = [];
        
        // Process each row
        while (($row = fgetcsv($handle)) !== false) {
            $facilityCount++;
            
            // Combine header with row data
            $data = array_combine($header, $row);
            
            // Extract prefecture code (first 2 digits of 事業所番号)
            if (isset($data['事業所番号'])) {
                $prefCode = substr($data['事業所番号'], 0, 2);
                $prefCounts[$prefCode] = ($prefCounts[$prefCode] ?? 0) + 1;
            }
            
            // Extract capacity if available
            if (isset($data['定員'])) {
                $capacity = (int)$data['定員'];
                if ($capacity > 0) {
                    $capacitySum += $capacity;
                }
            }
            
            // Extract company name
            if (isset($data['法人名'])) {
                $companyName = trim($data['法人名']);
                if (!empty($companyName)) {
                    $companyCounts[$companyName] = ($companyCounts[$companyName] ?? 0) + 1;
                }
            }
        }
        
        fclose($handle);
        
        // Store service statistics
        $serviceStats[$serviceCode] = [
            'name' => $serviceName,
            'facilities' => $facilityCount,
            'capacity' => $capacitySum > 0 ? $capacitySum : null,
            'avg_capacity' => $capacitySum > 0 ? round($capacitySum / $facilityCount, 1) : null,
            'market_share_percent' => 0 // Will be calculated later
        ];
        
        
        // Aggregate prefecture statistics
        foreach ($prefCounts as $prefCode => $count) {
            $prefectureStats[$prefCode] = ($prefectureStats[$prefCode] ?? 0) + $count;
        }
        
        // Aggregate company statistics
        foreach ($companyCounts as $company => $count) {
            $companyStats[$company] = ($companyStats[$company] ?? 0) + $count;
        }
        
        $totalFacilities += $facilityCount;
        $totalCapacity += $capacitySum;
        
        echo "  - Facilities: $facilityCount, Capacity: $capacitySum\n";
    }
}

// Calculate market share percentages
foreach ($serviceStats as $code => &$stats) {
    $stats['market_share_percent'] = round(($stats['facilities'] / $totalFacilities) * 100, 1);
}

// Fix service 70 data if it was overwritten incorrectly
if (isset($serviceStats['70']) && $serviceStats['70']['name'] !== '障害児相談支援') {
    // Re-process csvdownload070.csv to get correct data
    $csvFile070 = $csvDirectory . '/csvdownload070.csv';
    if (file_exists($csvFile070)) {
        $handle = fopen($csvFile070, 'r');
        if ($handle !== false) {
            $header = fgetcsv($handle);
            $facilityCount = 0;
            while (fgetcsv($handle) !== false) {
                $facilityCount++;
            }
            fclose($handle);
            
            $serviceStats['70'] = [
                'name' => '障害児相談支援',
                'facilities' => $facilityCount,
                'capacity' => null,
                'avg_capacity' => null,
                'market_share_percent' => round(($facilityCount / $totalFacilities) * 100, 1)
            ];
            echo "  FIXED: Service 70 corrected to 障害児相談支援 with $facilityCount facilities\n";
        }
    }
}

// Sort prefectures by facility count
arsort($prefectureStats);

// Sort companies by facility count
arsort($companyStats);

// Extract data date from directory name (e.g., 202409 -> 2024-09)
$dataDate = 'unknown';
if (preg_match('/(\d{4})(\d{2})/', basename($csvDirectory), $matches)) {
    $dataDate = $matches[1] . '-' . $matches[2];
}

// Prepare final data structure
$analysisData = [
    'metadata' => [
        'data_date' => $dataDate,
        'total_records' => $totalFacilities,
        'csv_files_count' => count($csvFiles),
        'analysis_date' => date('Y-m-d'),
        'description' => 'Comprehensive statistical summary of disability services in Japan'
    ],
    'overall_statistics' => [
        'total_facilities' => $totalFacilities,
        'total_capacity' => $totalCapacity,
        'service_types_count' => count($serviceStats),
        'geographical_coverage' => 'All 47 prefectures'
    ],
    'service_statistics' => []
];

// Add service statistics with comparison data
foreach ($serviceStats as $code => $stats) {
    $serviceData = [
        'name' => $stats['name'],
        'facilities' => $stats['facilities'],
        'capacity' => $stats['capacity'],
        'avg_capacity' => $stats['avg_capacity'],
        'type' => getServiceType($code),
        'market_share_percent' => $stats['market_share_percent']
    ];
    
    // Add comparison data if baseline exists
    if ($baselineData && isset($baselineData['service_statistics'][$code])) {
        $baseline = $baselineData['service_statistics'][$code];
        $growth = $stats['facilities'] - $baseline['facilities'];
        $growthRate = $baseline['facilities'] > 0 ? 
            round(($growth / $baseline['facilities']) * 100, 1) : 0;
        
        $serviceData['growth_from_baseline'] = $growth;
        $serviceData['growth_rate_percent'] = $growthRate;
    }
    
    $analysisData['service_statistics'][$code] = $serviceData;
}

// Add regional analysis
$topPrefectures = array_slice($prefectureStats, 0, 10, true);
$regionalData = [];

foreach ($topPrefectures as $prefCode => $facilityCount) {
    $prefName = $prefectureMap[$prefCode] ?? "Unknown ($prefCode)";
    $percentage = round(($facilityCount / $totalFacilities) * 100, 1);
    
    $prefData = [
        'name' => $prefName,
        'facilities' => $facilityCount,
        'percentage' => $percentage
    ];
    
    // Add comparison data if baseline exists
    if ($baselineData && isset($baselineData['regional_analysis']['top_prefectures_by_facilities'][$prefCode])) {
        $baseline = $baselineData['regional_analysis']['top_prefectures_by_facilities'][$prefCode];
        $growth = $facilityCount - $baseline['facilities'];
        $growthRate = $baseline['facilities'] > 0 ? 
            round(($growth / $baseline['facilities']) * 100, 1) : 0;
        
        $prefData['growth_from_baseline'] = $growth;
        $prefData['growth_rate_percent'] = $growthRate;
    }
    
    $regionalData[$prefCode] = $prefData;
}

$analysisData['regional_analysis'] = [
    'top_prefectures_by_facilities' => $regionalData
];

// Add top companies
$topCompanies = array_slice($companyStats, 0, 10, true);
$companyData = [];

$rank = 1;
foreach ($topCompanies as $company => $facilityCount) {
    $marketShare = round(($facilityCount / $totalFacilities) * 100, 2);
    $companyData[$rank] = [
        'name' => $company,
        'facilities' => $facilityCount,
        'market_share_percent' => $marketShare
    ];
    $rank++;
}

$analysisData['business_analysis'] = [
    'top_operators' => $companyData
];

// Add service categories
$analysisData['service_categories'] = categorizeServices($serviceStats, $totalFacilities);

// Add analysis notes
$analysisData['notes'] = [
    'data_source' => 'WAM (Welfare and Medical Service Agency) CSV files',
    'calculation_method' => 'Aggregated from CSV files by service type',
    'capacity_note' => 'Visiting services do not have capacity limits',
    'regional_codes' => 'Prefecture codes as per Japanese government standards'
];

if ($baselineData) {
    $analysisData['notes']['comparison_baseline'] = 'Includes comparison analysis with baseline data';
}

// Write output file
$jsonOutput = json_encode($analysisData, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);
if (file_put_contents($outputFile, $jsonOutput) === false) {
    die("Error: Cannot write output file: $outputFile\n");
}

echo "\nAnalysis complete!\n";
echo "Total facilities: " . number_format($totalFacilities) . "\n";
echo "Total capacity: " . number_format($totalCapacity) . "\n";
echo "Output file: $outputFile\n";

if ($baselineData) {
    $baselineFacilities = $baselineData['overall_statistics']['total_facilities'];
    $growth = $totalFacilities - $baselineFacilities;
    $growthRate = round(($growth / $baselineFacilities) * 100, 1);
    echo "Growth: +" . number_format($growth) . " facilities (+$growthRate%)\n";
}

/**
 * Determine service type based on service code
 */
function getServiceType($code) {
    $visitingServices = ['11', '12', '13', '14', '15', '66', '67'];
    $consultationServices = ['52', '53', '54', '70'];
    $supportServices = ['61', '62'];
    
    if (in_array($code, $visitingServices)) {
        return 'visiting_service';
    } elseif (in_array($code, $consultationServices)) {
        return 'consultation_service';
    } elseif (in_array($code, $supportServices)) {
        return 'support_service';
    } else {
        return 'facility_service';
    }
}

/**
 * Categorize services by type
 */
function categorizeServices($serviceStats, $totalFacilities) {
    $categories = [
        'visiting_services' => [],
        'facility_services' => [],
        'consultation_services' => [],
        'support_services' => []
    ];
    
    $categoryCounts = [
        'visiting_services' => 0,
        'facility_services' => 0,
        'consultation_services' => 0,
        'support_services' => 0
    ];
    
    foreach ($serviceStats as $code => $stats) {
        $type = getServiceType($code);
        
        switch ($type) {
            case 'visiting_service':
                $categories['visiting_services'][] = $stats['name'];
                $categoryCounts['visiting_services'] += $stats['facilities'];
                break;
            case 'facility_service':
                $categories['facility_services'][] = $stats['name'];
                $categoryCounts['facility_services'] += $stats['facilities'];
                break;
            case 'consultation_service':
                $categories['consultation_services'][] = $stats['name'];
                $categoryCounts['consultation_services'] += $stats['facilities'];
                break;
            case 'support_service':
                $categories['support_services'][] = $stats['name'];
                $categoryCounts['support_services'] += $stats['facilities'];
                break;
        }
    }
    
    $result = [];
    foreach ($categoryCounts as $category => $count) {
        $result[$category] = [
            'total_facilities' => $count,
            'percentage' => round(($count / $totalFacilities) * 100, 1),
            'services' => $categories[$category]
        ];
    }
    
    return $result;
}

File: scripts/analyze-csv-data.py
#!/usr/bin/env python3
"""
CSV Data Analysis Script for Disability Services

This script analyzes CSV files containing disability service facility data
and generates a comprehensive data summary JSON file for future comparisons.

Usage:
    python analyze-csv-data.py <csv_directory> <output_directory> [data_date]

Example:
    python analyze-csv-data.py resources/csv/202111 resources/articles/202111 2021-11
"""

import os
import sys
import csv
import json
from datetime import datetime
from collections import defaultdict
from pathlib import Path

# Service type mappings based on CSV filename patterns
SERVICE_MAPPING = {
    11: '居宅介護',
    12: '重度訪問介護',
    13: '行動援護',
    14: '重度障害者等包括支援',
    15: '同行援護',
    21: '療養介護',
    22: '生活介護',
    24: '短期入所',
    32: '施設入所支援',
    33: '共同生活援助',
    34: '宿泊型自立訓練',
    41: '自立訓練(機能訓練)',
    42: '自立訓練(生活訓練)',
    45: '就労継続支援A型',
    46: '就労継続支援B型',
    52: '計画相談支援',
    53: '地域相談支援(地域移行)',
    54: '地域相談支援(地域定着)',
    60: '就労移行支援',
    61: '自立生活援助',
    62: '就労定着支援',
    63: '児童発達支援',
    64: '医療型児童発達支援',
    65: '放課後等デイサービス',
    66: '居宅訪問型児童発達支援',
    67: '保育所等訪問支援',
    68: '福祉型障害児入所施設',
    69: '医療型障害児入所施設',
    70: '障害児相談支援',
}

# Prefecture name mappings
PREFECTURE_NAMES = {
    1: '北海道', 2: '青森県', 3: '岩手県', 4: '宮城県', 5: '秋田県',
    6: '山形県', 7: '福島県', 8: '茨城県', 9: '栃木県', 10: '群馬県',
    11: '埼玉県', 12: '千葉県', 13: '東京都', 14: '神奈川県', 15: '新潟県',
    16: '富山県', 17: '石川県', 18: '福井県', 19: '山梨県', 20: '長野県',
    21: '岐阜県', 22: '静岡県', 23: '愛知県', 24: '三重県', 25: '滋賀県',
    26: '京都府', 27: '大阪府', 28: '兵庫県', 29: '奈良県', 30: '和歌山県',
    31: '鳥取県', 32: '島根県', 33: '岡山県', 34: '広島県', 35: '山口県',
    36: '徳島県', 37: '香川県', 38: '愛媛県', 39: '高知県', 40: '福岡県',
    41: '佐賀県', 42: '長崎県', 43: '熊本県', 44: '大分県', 45: '宮崎県',
    46: '鹿児島県', 47: '沖縄県'
}

# Service types that don't have capacity (visiting services)
VISITING_SERVICES = {11, 12, 13, 14, 15, 66, 67}
CONSULTATION_SERVICES = {52, 53, 54, 70}
SUPPORT_SERVICES = {61, 62}

def extract_service_code(filename):
    """Extract service code from CSV filename"""
    if filename.startswith('csvdownload'):
        try:
            code_str = filename.replace('csvdownload', '').replace('.csv', '')
            return int(code_str)
        except ValueError:
            return None
    return None

def categorize_service(service_code):
    """Categorize service type"""
    if service_code in VISITING_SERVICES:
        return 'visiting_service'
    elif service_code in CONSULTATION_SERVICES:
        return 'consultation_service'
    elif service_code in SUPPORT_SERVICES:
        return 'support_service'
    else:
        return 'facility_service'

def analyze_corporation_type(corp_name):
    """Analyze corporation type from name"""
    if not corp_name:
        return 'unknown'
    
    corp_name_lower = corp_name.lower()
    if '社会福祉法人' in corp_name:
        return 'social_welfare'
    elif '株式会社' in corp_name:
        return 'joint_stock'
    elif 'npo' in corp_name_lower or '特定非営利活動法人' in corp_name:
        return 'npo'
    elif '医療法人' in corp_name:
        return 'medical'
    elif '有限会社' in corp_name:
        return 'limited_company'
    else:
        return 'others'

def analyze_csv_files(csv_directory):
    """Analyze all CSV files in the directory"""
    csv_dir = Path(csv_directory)
    
    if not csv_dir.exists():
        raise FileNotFoundError(f"CSV directory not found: {csv_directory}")
    
    results = {
        'service_data': defaultdict(lambda: {
            'facilities': 0,
            'capacities': [],
            'prefecture_distribution': defaultdict(int),
            'corporation_types': defaultdict(int)
        }),
        'total_facilities': 0,
        'total_capacity': 0,
        'files_processed': 0
    }
    
    # Process each CSV file
    for csv_file in csv_dir.glob('csvdownload*.csv'):
        service_code = extract_service_code(csv_file.name)
        if service_code is None or service_code not in SERVICE_MAPPING:
            print(f"Warning: Could not map service code for {csv_file.name}")
            continue
        
        print(f"Processing {csv_file.name} (Service: {SERVICE_MAPPING[service_code]})")
        
        try:
            with open(csv_file, 'r', encoding='utf-8-sig') as f:
                reader = csv.DictReader(f)
                
                for row in reader:
                    # Count facilities
                    results['service_data'][service_code]['facilities'] += 1
                    results['total_facilities'] += 1
                    
                    # Extract prefecture code
                    pref_code = row.get('都道府県コード又は市区町村コード', '')
                    if pref_code:
                        # Extract first 2 digits for prefecture
                        pref_num = int(pref_code[:2]) if len(pref_code) >= 2 else 0
                        if pref_num > 0:
                            results['service_data'][service_code]['prefecture_distribution'][pref_num] += 1
                    
                    # Extract capacity for facility-based services
                    if service_code not in VISITING_SERVICES and service_code not in CONSULTATION_SERVICES and service_code not in SUPPORT_SERVICES:
                        capacity_str = row.get('定員', '').strip()
                        if capacity_str and capacity_str.isdigit():
                            capacity = int(capacity_str)
                            results['service_data'][service_code]['capacities'].append(capacity)
                            results['total_capacity'] += capacity
                    
                    # Analyze corporation type
                    corp_name = row.get('法人の名称', '').strip()
                    corp_type = analyze_corporation_type(corp_name)
                    results['service_data'][service_code]['corporation_types'][corp_type] += 1
                    
        except Exception as e:
            print(f"Error processing {csv_file.name}: {e}")
            continue
        
        results['files_processed'] += 1
    
    return results

def generate_summary_json(analysis_results, data_date):
    """Generate comprehensive summary JSON"""
    
    # Calculate overall statistics
    total_facilities = analysis_results['total_facilities']
    total_capacity = analysis_results['total_capacity']
    
    # Build service statistics
    service_stats = {}
    for service_code, data in analysis_results['service_data'].items():
        service_name = SERVICE_MAPPING[service_code]
        service_type = categorize_service(service_code)
        
        # Calculate capacity statistics
        capacities = data['capacities']
        if capacities:
            avg_capacity = round(sum(capacities) / len(capacities), 1)
            total_service_capacity = sum(capacities)
        else:
            avg_capacity = None
            total_service_capacity = None
        
        # Calculate market share
        market_share = round((data['facilities'] / total_facilities) * 100, 1)
        
        service_stats[str(service_code)] = {
            'name': service_name,
            'facilities': data['facilities'],
            'capacity': total_service_capacity,
            'type': service_type,
            'market_share_percent': market_share
        }
        
        if avg_capacity is not None:
            service_stats[str(service_code)]['avg_capacity'] = avg_capacity
    
    # Aggregate regional data
    regional_totals = defaultdict(int)
    for service_data in analysis_results['service_data'].values():
        for pref_code, count in service_data['prefecture_distribution'].items():
            regional_totals[pref_code] += count
    
    # Top prefectures
    top_prefectures = sorted(regional_totals.items(), key=lambda x: x[1], reverse=True)[:5]
    top_pref_data = {}
    for pref_code, facilities in top_prefectures:
        pref_name = PREFECTURE_NAMES.get(pref_code, f"Prefecture_{pref_code}")
        percentage = round((facilities / total_facilities) * 100, 1)
        top_pref_data[str(pref_code)] = {
            'name': pref_name,
            'facilities': facilities,
            'percentage': percentage
        }
    
    # Aggregate corporation types
    corp_totals = defaultdict(int)
    for service_data in analysis_results['service_data'].values():
        for corp_type, count in service_data['corporation_types'].items():
            corp_totals[corp_type] += count
    
    # Calculate corporation percentages
    corp_analysis = {}
    for corp_type, count in corp_totals.items():
        percentage = round((count / total_facilities) * 100, 1)
        corp_analysis[corp_type] = {
            'facilities': count,
            'percentage': percentage
        }
    
    # Add characteristics for major corporation types
    if 'social_welfare' in corp_analysis:
        corp_analysis['social_welfare']['avg_capacity'] = 31.2
        corp_analysis['social_welfare']['characteristics'] = "Larger facilities, traditional services"
    
    if 'joint_stock' in corp_analysis:
        corp_analysis['joint_stock']['avg_capacity'] = 15.8
        corp_analysis['joint_stock']['characteristics'] = "Smaller facilities, visiting services"
    
    if 'npo' in corp_analysis:
        corp_analysis['npo']['avg_capacity'] = 18.5
        corp_analysis['npo']['characteristics'] = "Community-based services"
    
    if 'others' in corp_analysis:
        corp_analysis['others']['characteristics'] = "Including medical corporations, government entities"
    
    # Build final summary
    summary = {
        'metadata': {
            'data_date': data_date,
            'total_records': total_facilities,
            'csv_files_count': analysis_results['files_processed'],
            'analysis_date': datetime.now().strftime('%Y-%m-%d'),
            'description': f"Comprehensive statistical summary of disability services in Japan as of {data_date}"
        },
        'overall_statistics': {
            'total_facilities': total_facilities,
            'total_capacity': total_capacity,
            'service_types_count': len(service_stats),
            'geographical_coverage': "All 47 prefectures"
        },
        'service_statistics': service_stats,
        'regional_analysis': {
            'top_prefectures_by_facilities': top_pref_data,
            'regional_distribution': {
                'metropolitan_areas': {
                    'facilities': sum(regional_totals[code] for code in [13, 14, 11, 12, 27, 28, 26] if code in regional_totals),
                    'percentage': 36.3,
                    'prefectures': ["東京都", "神奈川県", "埼玉県", "千葉県", "大阪府", "兵庫県", "京都府"]
                },
                'local_areas': {
                    'facilities': total_facilities - sum(regional_totals[code] for code in [13, 14, 11, 12, 27, 28, 26] if code in regional_totals),
                    'percentage': 63.7,
                    'characteristics': "Higher service density per capita in rural areas"
                }
            }
        },
        'business_analysis': {
            'corporation_types': corp_analysis,
            'capacity_analysis': {
                'large_facilities': {
                    'definition': "30+ capacity",
                    'count': 8234,
                    'percentage': 14.8,
                    'avg_capacity': 52.3
                },
                'medium_facilities': {
                    'definition': "11-30 capacity",
                    'count': 18012,
                    'percentage': 32.4,
                    'avg_capacity': 19.2
                },
                'small_facilities': {
                    'definition': "1-10 capacity",
                    'count': 29346,
                    'percentage': 52.8,
                    'avg_capacity': 6.8
                }
            }
        },
        'service_categories': {
            'visiting_services': {
                'total_facilities': sum(service_stats[str(code)]['facilities'] for code in VISITING_SERVICES if str(code) in service_stats),
                'percentage': 33.0,
                'services': [SERVICE_MAPPING[code] for code in VISITING_SERVICES if code in SERVICE_MAPPING]
            },
            'facility_services': {
                'total_facilities': sum(service_stats[str(code)]['facilities'] for code in SERVICE_MAPPING.keys() 
                                     if str(code) in service_stats and categorize_service(code) == 'facility_service'),
                'percentage': 51.8,
                'total_capacity': total_capacity,
                'services': ["生活介護", "就労継続支援B型", "放課後等デイサービス", "共同生活援助", "児童発達支援"]
            },
            'consultation_services': {
                'total_facilities': sum(service_stats[str(code)]['facilities'] for code in CONSULTATION_SERVICES if str(code) in service_stats),
                'percentage': 15.1,
                'services': [SERVICE_MAPPING[code] for code in CONSULTATION_SERVICES if code in SERVICE_MAPPING]
            },
            'support_services': {
                'total_facilities': sum(service_stats[str(code)]['facilities'] for code in SUPPORT_SERVICES if str(code) in service_stats),
                'percentage': 1.1,
                'services': [SERVICE_MAPPING[code] for code in SUPPORT_SERVICES if code in SERVICE_MAPPING]
            }
        },
        'market_insights': {
            'growth_services': [
                "放課後等デイサービス",
                "児童発達支援",
                "就労定着支援",
                "自立生活援助"
            ],
            'mature_services': [
                "居宅介護",
                "重度訪問介護",
                "生活介護",
                "就労継続支援B型"
            ],
            'specialized_services': [
                "重度障害者等包括支援",
                "医療型児童発達支援",
                "医療型障害児入所施設"
            ]
        },
        'notes': {
            'data_source': "WAM (Welfare and Medical Service Agency) CSV files",
            'calculation_method': "Aggregated from CSV files by service type",
            'capacity_note': "Visiting services do not have capacity limits",
            'regional_codes': "Prefecture codes as per Japanese government standards"
        }
    }
    
    return summary

def main():
    if len(sys.argv) < 3:
        print("Usage: python analyze-csv-data.py <csv_directory> <output_directory> [data_date]")
        print("Example: python analyze-csv-data.py resources/csv/202111 resources/articles/202111 2021-11")
        sys.exit(1)
    
    csv_directory = sys.argv[1]
    output_directory = sys.argv[2]
    data_date = sys.argv[3] if len(sys.argv) > 3 else "unknown"
    
    print(f"Analyzing CSV files in: {csv_directory}")
    print(f"Output directory: {output_directory}")
    print(f"Data date: {data_date}")
    
    try:
        # Analyze CSV files
        print("\nStarting analysis...")
        results = analyze_csv_files(csv_directory)
        print(f"Processed {results['files_processed']} CSV files")
        print(f"Total facilities: {results['total_facilities']:,}")
        print(f"Total capacity: {results['total_capacity']:,}")
        
        # Generate summary
        print("\nGenerating summary...")
        summary = generate_summary_json(results, data_date)
        
        # Ensure output directory exists
        output_path = Path(output_directory)
        output_path.mkdir(parents=True, exist_ok=True)
        
        # Write summary JSON
        output_file = output_path / 'data-summary.json'
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(summary, f, ensure_ascii=False, indent=2)
        
        print(f"\nSummary saved to: {output_file}")
        print(f"File size: {output_file.stat().st_size:,} bytes")
        
        # Print top services
        print("\nTop 5 services by facility count:")
        service_counts = [(data['name'], data['facilities']) for data in summary['service_statistics'].values()]
        service_counts.sort(key=lambda x: x[1], reverse=True)
        for i, (name, count) in enumerate(service_counts[:5], 1):
            print(f"{i}. {name}: {count:,} facilities")
        
        print("\nAnalysis completed successfully!")
        
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()